### PR TITLE
DRILL-6524: Prevent incorrect scalar replacement for the case of assigning references inside if block

### PR DESCRIPTION
**Pull request by vvysotskyi**
This pull request contains two commits:
- The first one contains changes to prevent incorrect scalar replacement for the case of assigning references inside if block to avoid wrong results for some cases
- The second commit contains changes to assign holder fields instead of assigning just object reference for the case when the source object is not used anymore (a common example is the result of UDF execution), so it will allow producing scalar replacement more frequently.

Therefore with these two commits, a scalar replacement will be produced correctly and more frequently despite preventing it in the first commit.

For detailed problem description please see comment in [DRILL-6524](https://issues.apache.org/jira/browse/DRILL-6524)

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@vdiravka could you please review?

File: exec/java-exec/src/test/java/org/apache/drill/exec/compile/TestClassTransformation.java [hunk 1]
```diff
@@ -79,11 +86,10 @@ public void testCompilationNoDebug() throws CompileException, ClassNotFoundExcep
     ClassSet classSet = new ClassSet(null, cg.getDefinition().getTemplateClassName(), cg.getMaterializedClassName());
     String sourceCode = cg.generateAndGet();
     sessionOptions.setLocalOption(ClassCompilerSelector.JAVA_COMPILER_OPTION, ClassCompilerSelector.CompilerPolicy.JDK.name());
-
     sessionOptions.setLocalOption(ClassCompilerSelector.JAVA_COMPILER_DEBUG_OPTION, false);
-    @SuppressWarnings("resource")
```
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/exec/compile/TestClassTransformation.java]: Is this warning displayed now?

File: exec/vector/src/main/java/org/apache/drill/exec/vector/ValueHolderHelper.java [hunk 2]
```diff
@@ -229,4 +231,77 @@ public static VarDecimalHolder getVarDecimalHolder(DrillBuf buf, BigDecimal bigD
...
+          case VAR16CHAR:
+          case VARBINARY:
+          case VARCHAR:
+            result.add("start");
+            result.add("end");
+            result.add("buffer");
+            return result;
+          case UNION:
+            result.add("reader");
+            return result;
```
Comment (vdiravka) [inline exec/vector/src/main/java/org/apache/drill/exec/vector/ValueHolderHelper.java]: What about additional datatypes: `MAP`, `LIST`, `NULL` ( and others from `Types.getBaseSqlTypeName()`)?
Is `emptyResult` fine for them?

File: exec/java-exec/src/main/java/org/apache/drill/exec/compile/bytecode/MethodAnalyzer.java [hunk 3]
```diff
@@ -36,70 +46,147 @@
  * as factories that will provide our own derivative of Frame<> which we use to detect
  */
 public class MethodAnalyzer<V extends Value> extends Analyzer <V> {
+
+  // list of method instructions which is analyzed
+  private InsnList insnList;
+
+  /**
+   * Constructor.
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/compile/bytecode/MethodAnalyzer.java]: Thanks, removed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/compile/bytecode/MethodAnalyzer.java [hunk 4]
```diff
@@ -36,70 +46,147 @@
...
-   * that are assigned to the same local variable slot.
+   * that are assigned to the same local variable slot. Also it controls stack to determine whether
+   * object was assigned to the value declared outside of conditional block.
    *
    * <p>Since this is almost a pass-through, the constructors' arguments match
    * those from Frame<>.
    */
   private static class AssignmentTrackingFrame<V extends Value> extends Frame<V> {
+
+    // represents stacks of variable sets declared inside current code block
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/compile/bytecode/MethodAnalyzer.java]: Thanks, done.

File: exec/java-exec/src/test/java/org/apache/drill/exec/compile/TestClassTransformation.java (hunk 1 above)
Comment (vvysotskyi) [reply exec/java-exec/src/test/java/org/apache/drill/exec/compile/TestClassTransformation.java]: No, it is not displayed, after removing this line.

File: exec/vector/src/main/java/org/apache/drill/exec/vector/ValueHolderHelper.java (hunk 2 above)
Comment (vvysotskyi) [reply exec/vector/src/main/java/org/apache/drill/exec/vector/ValueHolderHelper.java]: They are handled separately in the code generation classes, so there is no need to specify them here.

**Pull request closed without merging**